Add extra sensors for BMW ConnectedDrive

gerard33 opened the pull request:
## Description:
The following items are added to BMW ConnectedDrive:
- Binary sensors for `Doors`, `Windows` and `Door lock state`
- Lock

For the existing sensors `remaining_range_fuel`, `mileage` and `remaining_fuel` the attribute `car: <car_model>` and an applicable mdi icon are added.

~~**Breaking change**~~ reverted in #13380 

----
**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4733

## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

gerard33 committed:
Added extra sensors for BMW ConnectedDrive

gerard33 commented:
@ChristianKuehnel I have added the binary sensor, lock and made some small changes to the existing sensor. Can you please review these?

MartinHjelmare commented on homeassistant/components/binary_sensor/bmw_connected_drive.py:
```diff
@@ -0,0 +1,123 @@
+        return self._state
+
+    @property
+    def unit_of_measurement(self) -> str:
```
Binary sensors don't have a unit of measurement.

gerard33 replied:
Removed it.

MartinHjelmare commented on homeassistant/components/binary_sensor/bmw_connected_drive.py:
```diff
@@ -0,0 +1,123 @@
+https://home-assistant.io/components/binary_sensor.bmw_connected_drive/
+"""
+import logging
+import asyncio
```
Please sort 🔤 within groups standard library, 3rd party and homeassistant.

gerard33 replied:
Done

MartinHjelmare commented on homeassistant/components/binary_sensor/bmw_connected_drive.py:
```diff
@@ -0,0 +1,123 @@
+        """Return the state attributes of the binary sensor."""
+        vehicle_state = self._vehicle.state
+        result = {
+            'last_update': vehicle_state.timestamp,
```
Time stamp is not allowed in state attributes. The state should not update just because of a time change.

gerard33 replied:
This shows the last update time which is received from the car via the bmw_connected_drive module.
This time stamp is only updated together with other updates received from the car, e.g. an update on the mileage, see the screenshot below.

It's the same usage as in the [Mercedes ME](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/binary_sensor/mercedesme.py#L57) component.

![screenshot_7](https://user-images.githubusercontent.com/11230573/36565528-003e73c6-1821-11e8-87f9-09e7fa7f66c4.png)

ChristianKuehnel replied:
@gerard33 
I get your point, I agree it's quite useful to see when the last measurement was taken. But isn't the information redundant? You would also get the age of the information from the headline "3 hours ago".

MartinHjelmare replied:
Yes, this is redundant, please remove it.

gerard33 replied:
As a matter of fact it is not redudant. The last update contains the time on which the update from the car is received, but that doesn't always mean that every sensor is being updated. So this can result in a different time shown at the time information in the headline than at the last update time. See the example with the below screenshots.
With this update the door lock state has changed, but the windows state hasn't, resulting in different times for the windows sensor in the headline versus the last update time.

I think it has added value to show this attribute as you can easily see if the sensor has changed or not as part of the last update received from the car.

![img_8122_2](https://user-images.githubusercontent.com/11230573/37354725-9a629782-26e2-11e8-89d8-547cd253aecd.png)
![img_8123_2](https://user-images.githubusercontent.com/11230573/37354726-9a7c0ae6-26e2-11e8-93fd-589ae6637a9f.png)

MartinHjelmare replied:
If a state attribute changes, the last_updated attribute of the state will be updated with the time, so this information is already available in the the state machine. If a user wants to know about this we offer the template sensor which can display state attributes including last_updated. We shouldn't duplicate this information in the state machine.

MartinHjelmare commented on homeassistant/components/binary_sensor/bmw_connected_drive.py:
```diff
@@ -0,0 +1,123 @@
+
+        if self._attribute == 'all_lids_closed':
+            for lid in vehicle_state.lids:
+                result[lid.name] = lid.state.value
```
Is lid name lowercase snakecase? That's the home assistant standard.

gerard33 replied:
It is, this is an example of a lid name: `door_driver_front`.

MartinHjelmare replied:
:+1:

MartinHjelmare commented on homeassistant/components/binary_sensor/bmw_connected_drive.py:
```diff
@@ -0,0 +1,123 @@
+            for lid in vehicle_state.lids:
+                result[lid.name] = lid.state.value
+        elif self._attribute == 'all_windows_closed':
+            for lid in vehicle_state.windows:
```
This is iterating windows not lids.

gerard33 replied:
Changed it to window :)

MartinHjelmare commented on homeassistant/components/binary_sensor/bmw_connected_drive.py:
```diff
@@ -0,0 +1,123 @@
+        self._name = sensor_name
+        self._device_class = device_class
+        self._state = None
+        self._unit_of_measurement = None
```
Remove this.

gerard33 replied:
Done

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,122 @@
+https://home-assistant.io/components/lock.bmw_connected_drive/
+"""
+import logging
+import asyncio
```
Please sort 🔤.

gerard33 replied:
Done

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,122 @@
+        """Return the state attributes of the lock."""
+        vehicle_state = self._vehicle.state
+        return {
+            'last_update': vehicle_state.timestamp,
```
Please remove.

gerard33 replied:
See my comment above an this last update time.

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,122 @@
+        """Lock the car."""
+        from bimmer_connected.remote_services import ExecutionState
+
+        _LOGGER.info("%s: locking doors", self._vehicle.modelName)
```
Debug at most.

gerard33 replied:
Changed to debug

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,122 @@
+        return self._state == STATE_LOCKED
+
+    @asyncio.coroutine
+    def async_lock(self, **kwargs):
```
Use non async version of the method. The interface library is not async.

gerard33 replied:
Changed it.
@ChristianKuehnel can you please also check the current solution?

ChristianKuehnel replied:
@MartinHjelmare 
Here we basically need to poll the server until we get the proper answer back. And that might take 10 to 30 seconds.

So my assumption here is: the locking a thread process for that long is a bad idea. Thus I would have made this async. So some other coroutine can be executed while waiting.

What we could do:
Move all of the logic for awaiting a proper answer from the server to the library and then provide a sync and an async method that will only return once we have reached a final state...

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,122 @@
+        _LOGGER.info("%s: locking doors", self._vehicle.modelName)
+        status = self._vehicle.remote_services.trigger_remote_door_lock()
+        self._state = STATE_LOCKED
+        self.async_schedule_update_ha_state()
```
Since we register a callback for state updates, I expect that the callback is called when the state has changed on the device, (car lock). Isn't that true? If that is true, the lock method should only call the device lock method.

gerard33 replied:
Locking/unlocking the car doesn't trigger a state update.

ChristianKuehnel replied:
We're polling the update every 5 minutes. So if we do not change the state here, the UI will show an old state for 5 minutes.

MartinHjelmare replied:
I think a middle way is to do an optimistic state update immediately in home assistant, then have the library call the update callback after 10 seconds with the correct state when we are sure that that is present.

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,122 @@
+        self.hass.async_add_job(self._account.update)
+
+    @asyncio.coroutine
+    def async_unlock(self, **kwargs):
```
See above.

gerard33 replied:
Done

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,122 @@
+        _LOGGER.info("%s: unlocking doors", self._vehicle.modelName)
+        status = self._vehicle.remote_services.trigger_remote_door_unlock()
+        self._state = STATE_UNLOCKED
+        self.async_schedule_update_ha_state()
```
See above.

gerard33 replied:
Locking/unlocking the car doesn't trigger a state update.

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,122 @@
+
+    def update(self):
+        """Update state of the lock."""
+        _LOGGER.info("%s: updating data for %s", self._vehicle.modelName,
```
Debug at most.

gerard33 replied:
Changed to debug

MartinHjelmare commented on homeassistant/components/sensor/bmw_connected_drive.py:
```diff
@@ -74,6 +83,16 @@ def unit_of_measurement(self) -> str:
+        vehicle_state = self._vehicle.state
+
+        return {
+            'last_update': vehicle_state.timestamp,
```
Please remove.

gerard33 replied:
See my comment above an this last update time.

gerard33 committed:
Updates based on review of @MartinHjelmare

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,119 @@
+        _LOGGER.debug("%s: locking doors", self._vehicle.modelName)
+        status = self._vehicle.remote_services.trigger_remote_door_lock()
+        self._state = STATE_LOCKED
+        while status.state != ExecutionState.EXECUTED:
```
This should probably be handled in the interface library.

gerard33 replied:
I have removed this check completly in here and will check with @ChristianKuehnel to handle this in the interface library.

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,119 @@
+        self._state = STATE_LOCKED
+        while status.state != ExecutionState.EXECUTED:
+            status = self._vehicle.remote_services.get_remote_service_status()
+            time.sleep(1)
```
Sleeping is not allowed in the executor thread pool.

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,119 @@
+
+        _LOGGER.debug("%s: locking doors", self._vehicle.modelName)
+        status = self._vehicle.remote_services.trigger_remote_door_lock()
+        self._state = STATE_LOCKED
```
Either the update callback should handle setting state or we should do an optimistic state update from here, immediately. The logic below doesn't belong in this method.

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,119 @@
+            time.sleep(1)
+        _LOGGER.debug("%s: doors locked", self._vehicle.modelName)
+        time.sleep(10)
+        self.hass.async_add_job(self._account.update)
```
Don't use async methods from a sync context.

gerard33 replied:
Removed

MartinHjelmare reviewed (commented):
It would be best if the interface library made sure the update callback got called upon new state after the lock is locked/unlocked.

gerard33 committed:
Updates based on 2nd review of @MartinHjelmare

ChristianKuehnel commented on homeassistant/components/binary_sensor/bmw_connected_drive.py:
```diff
@@ -0,0 +1,117 @@
+        self._account = account
+        self._vehicle = vehicle
+        self._attribute = attribute
+        self._name = sensor_name
```
Does the name of the sensor also contain the name of the vehicle?
If someone has more than one vehicle, we need to make sure that the sensors then get different names...

gerard33 replied:
The vehicle name is not show in the sensor name, but it is shown in the attributes. Not sure what is the best approach, but maybe those people with more than one vehicle can update the friendly name of the sensor? Because in my opinion it's not necessary to show the vehicle name in the sensor name if you have one BMW (which will be the majority I guess).

ChristianKuehnel replied:
Should we add an option where people can give names to their vehicles? Something like:

```yaml
bmw_connected_drive:
  my_account:
    username: USERNAME_BMW_CONNECTED_DRIVE
    password: PASSWORD_BMW_CONNECTED_DRIVE
    country: COUNTRY_BMW_CONNECTED_DRIVE
    vehicles:
    - vin: SOMEVIN
      name: his
    - vin: SOMEOTHERVIN
      name: hers
```

That way people can
* select which cars should be listed in HASS at all
* set custom names for their vehicles
* we have different names for the different vehicles?

gerard33 replied:
That is indeed something we could add as an option.

ChristianKuehnel committed:
Changed control flow for updates to support updates triggered by remote services.

* updated library version number

ChristianKuehnel committed:
Changed order of commands so that the UI looks consistent.

State of lock is now set optimisitcally before getting proper update
from the server. So that the state does not toggle in the UI.

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -71,14 +71,16 @@ def is_locked(self):
         """Lock the car."""
         _LOGGER.debug("%s: locking doors", self._vehicle.modelName)
-        self._vehicle.remote_services.trigger_remote_door_lock()
         self._state = STATE_LOCKED
```
Please add a comment before this, saying why we need a optimistic state update, when we also have a update callback.

gerard33 replied:
Added the comments

gerard33 committed:
Added comment on optimistic state

gerard33 committed:
Updated requirements_all.txt

gerard33 commented:
Closing and opening the PR to trigger another Travis CI check.

MartinHjelmare commented:
You changed access permissions for the requirements_all.txt file. Please revert that. That file should only be modified via the `gen_requirements_all.py` script.

gerard33 committed:
Revert access permission changes

gerard33 commented:
Another close/open for Travis CI.

gerard33 commented:
Travis CI gives some errors with `tests.components.test_pilight.TestPilight`.
Not sure what to do about that?

ChristianKuehnel commented on homeassistant/components/bmw_connected_drive.py:
```diff
@@ -16,7 +16,7 @@
 )
 
-REQUIREMENTS = ['bimmer_connected==0.3.0']
+REQUIREMENTS = ['bimmer_connected==0.4.0']
```
Please update to version `0.4.1`. I fixed some exceptions there and it's currently already in use on master

gerard33 replied:
Done

ChristianKuehnel commented:
@gerard33 Looks good to me.

As I updated to bimmer_connected version 0.4.1 on master you probably need to rebase your branch and re-run ./scripcs/gen-requirements.

Sorry for that, but I wanted to get the bugfix out quickly...

gerard33 committed:
Fix for Travis

gerard33 commented:
@ChristianKuehnel, I updated to bimmer_connected version 0.4.1.

MartinHjelmare commented on homeassistant/components/binary_sensor/bmw_connected_drive.py:
```diff
@@ -0,0 +1,117 @@
+            self._state = bool(vehicle_state.door_lock_state.value
+                               in ('SELECTIVELOCKED', 'UNLOCKED'))
+
+        self.schedule_update_ha_state()
```
It would be more efficient if you move this to a separate new method which calls this with argument `True` which will make sure `update` is called before the state is updated.

Then register this new method as state update callback in `async_added_to_hass`.

```py
def update_callback(self):
    """Schedule a state update."""
    self.schedule_update_ha_state(True)
```

gerard33 replied:
Updated this, can you please doublecheck if it's okay?

MartinHjelmare commented on homeassistant/components/binary_sensor/bmw_connected_drive.py:
```diff
@@ -0,0 +1,117 @@
+
+        Show latest data after startup.
+        """
+        self._account.add_update_listener(self.update)
```
```py
self._account.add_update_listener(
    self.update_callback)
```

gerard33 replied:
Done

MartinHjelmare commented on homeassistant/components/binary_sensor/bmw_connected_drive.py:
```diff
@@ -0,0 +1,117 @@
+        Show latest data after startup.
+        """
+        self._account.add_update_listener(self.update)
+        yield from self.hass.async_add_job(self.update)
```
Remove this and instead add an argument `True` to `add_devices`.

gerard33 replied:
Done

MartinHjelmare commented on homeassistant/components/binary_sensor/bmw_connected_drive.py:
```diff
@@ -0,0 +1,117 @@
+                device = BMWConnectedDriveSensor(account, vehicle, key,
+                                                 value[0], value[1])
+                devices.append(device)
+    add_devices(devices)
```
```py
add_devices(devices, True)
```
This will make sure `update` is called before adding the device.

gerard33 replied:
Done

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,108 @@
+        for vehicle in account.account.vehicles:
+            device = BMWLock(account, vehicle, 'lock', 'BMW lock')
+            devices.append(device)
+    add_devices(devices)
```
See above.

gerard33 replied:
Done

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,108 @@
+        self._state = (STATE_LOCKED if vehicle_state.door_lock_state.value
+                       in ('LOCKED', 'SECURED') else STATE_UNLOCKED)
+
+        self.schedule_update_ha_state()
```
See above.

gerard33 replied:
Done

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,108 @@
+
+        Show latest data after startup.
+        """
+        self._account.add_update_listener(self.update)
```
See above.

gerard33 replied:
Done

MartinHjelmare commented on homeassistant/components/lock/bmw_connected_drive.py:
```diff
@@ -0,0 +1,108 @@
+        Show latest data after startup.
+        """
+        self._account.add_update_listener(self.update)
+        yield from self.hass.async_add_job(self.update)
```
See above.

gerard33 replied:
Done

MartinHjelmare commented on homeassistant/components/sensor/bmw_connected_drive.py:
```diff
@@ -74,6 +83,16 @@ def unit_of_measurement(self) -> str:
+
     def update(self) -> None:
         """Read new state data from the library."""
         _LOGGER.debug('Updating %s', self.entity_id)
```
Replace this entity_id with name as in the other platforms.

Also please update the state update logic below as in the other platforms.

gerard33 replied:
Done

MartinHjelmare reviewed (changes requested):
Please update the state update logic in all platforms. It will be more efficient and more on line with the rest of home assistant.

Also please remove the time stamp from all state attributes. It's redundant.

With that done this should be good to go.

gerard33 committed:
Changes based on review by @MartinHjelmare

MartinHjelmare reviewed (approved):
Looks great!

MartinHjelmare commented:
Please add a paragraph and highlight the breaking changes in the description, so it's easy to copy paste that to the release notes.

gerard33 commented:
@MartinHjelmare thanks!
I have added a paragraph in the description highlighting the breaking changes to the existing sensors.

MartinHjelmare commented:
Great!

gerard33 commented:
@MartinHjelmare 
cc @ChristianKuehnel 

I just realized that the breaking change for the names of the sensors can be avoided.
I can leave the entity name of the sensor untouched and just add the new (more descriptive) name as friendly_name attribute to the sensors, so that name will be shown in the frontend together with the new mdi icons.

What do you think? If you agree I can make a PR (including the new binary sensors and lock) asap so this can be part of the HA 0.66 release.
Just let me know.

MartinHjelmare commented:
That won't work. We're not allowed to set `friendly_name`. What you can do is to set the entity_id. But I suggest you don't do that. Instead it would be better if we can define unique ids for each entity. Then the user can change entity_id if they want. Can we use the vin number?

gerard33 commented:
Ah, I see.
Yes the VIN is available both in the current API as in the new API #13305.
I suggest adding the VIN to the entity_id in the new PR then?
And for now change the sensor name back to the [currently used name](https://github.com/home-assistant/home-assistant/pull/12591/files#diff-47a1033d931d4fe7cfccdf789ef66559L51) to avoid having 2 breaking changes for this?

MartinHjelmare commented:
We should define `unique_id` property and use the vin number plus sensor type when applicable. Don't change entity_id.

gerard33 commented:
Ok. And shall I adapt this PR so there is no breaking change in the name of the current 3 sensors?

Then the `unique_id` can be implemented in the WIP PR.

MartinHjelmare commented:
If the breaking name change can be avoided, that sounds good.

Try to only have one new feature per PR.
